DOC-3498: tinymceai on-premises documentation (#4142)
tinymce/tinymce-docs: modules/ROOT/pages/tinymceai-on-premises.adoc

= TinyMCE AI on-premises
:navtitle: Overview
:description: Self-hosted AI writing assistance for TinyMCE; architecture, prerequisites, and topic guide
:keywords: AI, on-premises, self-hosted, deployment, overview

The TinyMCE AI on-premises service is a self-hosted back end that powers AI writing assistance. It can be used with the TinyMCE rich text editor, particularly the xref:tinymceai.adoc[TinyMCE AI plugin], or as a standalone service. It runs entirely within the host infrastructure. Document content, conversation history, file attachments, and user data stay within the host network and are not stored by Tiny. Data sent to a configured LLM provider is subject to that provider's data handling policies.

The service ships as a single Open Container Initiative (OCI) container image (`registry.containers.tiny.cloud/ai-service-tiny`). It exposes a REST API, a Management Panel, Server-Sent Events streaming, and an OpenAPI spec.

== Architecture

> 

Reviewer: per other comment, realised we don't actually describe the architecture anywhere. throwing this in from my old docs draft - feel free to massage

```suggestion

The TinyMCE AI service infrastructure consists of two layers:

The **application layer** runs the AI service and communicates with LLM providers to process AI requests. It may consist of one or more application server instances, each running the TinyMCE AI service Docker image, and, optionally, a load balancer to distribute incoming requests across the instances. The round-robin algorithm is recommended if using a load balancer.

The **data layer** consists of a SQL database, a Redis instance, and file storage. The purpose of each is:

- **SQL database:** Stores persistent data such as configurations, conversations, files, and documents.
- **Redis:** Handles temporary data and inter-instance communication for scaling. When multiple application instances are running, Redis ensures that data is shared correctly across all of them.
- **File storage:** Stores uploaded files and documents.

=== Data flow

```

Author: Incorporated — the Overview page now has a full architecture section describing three layers (client, application, data) with a deployment architecture diagram. Used your draft as the base and expanded with the client layer and a Mermaid diagram. See e006b46.

The service ships as a single Open Container Initiative (OCI) container image (`registry.containers.tiny.cloud/ai-service-tiny`). It exposes a REST API, a Management Panel, Server-Sent Events streaming, and an OpenAPI spec.

== Architecture

The infrastructure consists of three layers:

* The *client layer* runs in the browser (TinyMCE editor with the `tinymceai` plugin) and in the application back end (token endpoint that signs JWTs).
* The *application layer* runs the AI service container and communicates with LLM providers to process AI requests. It may consist of one or more instances behind a load balancer (round-robin recommended). Each instance runs the same stateless container image.
* The *data layer* consists of a SQL database, a Redis instance, and file storage:

> ** *SQL database*: stores persistent data: configurations, conversations, files, and documents.

Reviewer 2: ```suggestion
** *SQL database*: stores persistent data such as configuration and conversations
```

Does the AI service also users info here?

Author: Applied the shorter phrasing — now reads: "stores persistent data such as configurations and conversations."

Re: user info — the AI service stores a `sub` claim (user identifier from the JWT) to isolate conversations per-user, but does not store user profiles or credentials. Files/documents are stored in file storage (S3/Azure/filesystem), not the SQL database, so those are correctly omitted.

> ** *Redis*: handles temporary data and inter-instance communication for scaling. When multiple instances are running, Redis ensures data is shared correctly across all of them.

Reviewer 2: ```suggestion
** *Redis*: handles temporary data and inter-instance communication between multiple application instances at scale.
```

Author: Updated — aligned with the diagram label ('cache, coordination') and the dedicated Redis section. Current wording: "caching and coordination (SSE delivery, rate limits, pub/sub). Enables the AI service to remain stateless." More specific than the original, slightly different from the suggestion but technically accurate.

** *File storage*: stores uploaded files and documents (S3, Azure Blob, filesystem, or the database itself).

[.text-center]
image::tinymceai-on-premises/overview-fig-2.svg[alt="Deployment architecture showing data layer, application layer, client layer, and LLM provider connections",width=100%]

=== Data flow

> Data flow for a single AI request:

Reviewer 2: ```suggestion
AI Request Flow
```

Author: Kept as "Data flow" — it's broader and matches enterprise documentation patterns (CockroachDB, etc.). The introductory text already specifies "for a single AI request" so the context is clear.

> . The client application requests a JWT from the token endpoint.

Reviewer 2: ```suggestion
. The client application makes a request to retrieve a signed JWT from the token endpoint.
```

Author: Kept the shorter form — "The client application requests a JWT from the token endpoint" conveys the same meaning in fewer words. The diagram now also shows this with the arrow direction (App → Token).

. The client sends a prompt with the JWT to the AI service over HTTPS.
. The AI service verifies the token, checks per-feature permissions, and forwards the prompt to the configured large language model (LLM).
. The LLM streams its response back to the AI service.

> . The AI service relays the response to the client through Server-Sent Events (SSE).

Reviewer 2: So the AI services are just a gate keeper that validates permissions and forwards prompts to a LLM? I believe it could be re-worded to be simpler

Author: Good point — simplified step 3 to "The AI service verifies the token and forwards the prompt to the configured large language model (LLM)." Removes the permissions detail from the high-level flow. See 29e0a1c.

When used with TinyMCE `tinymceai`, the plugin handles steps 1, 2, and 5 automatically through the `tinymceai_token_provider` callback.

> IMPORTANT: The browser connects directly to the AI service — requests do not pass through the application back end. The AI service must be network-reachable from the end-user browser, which means it must have a public URL (or be accessible through a VPN/internal network when deployed on an intranet). Configure xref:tinymceai-on-premises-frameworks.adoc#_cross_origin_requests_to_the_ai_service[CORS] and xref:tinymceai-on-premises-production.adoc#_tls_https[TLS] on the AI service accordingly.

Reviewer 2: Maybe just a note instead of important mark

Author: Applied — changed to NOTE. See 29e0a1c.

> The shared secret (API Secret) never leaves the back end; the editor and the AI service only ever see signed tokens.

Reviewer 2: I believe that the secret never leaves the AI service Docker container.

Author: Good catch — the secret is a shared symmetric key (HS256), so it exists in both the token endpoint AND the AI service container. Reworded to: "The shared secret (API Secret) exists only in the application back end (token endpoint) and the AI service container. The browser and the editor never see it — they only handle signed tokens." See 29e0a1c.

|The service is stateless; add replicas behind a load balancer without shared local state.

|OpenAPI specification
|Published at `/v1/api/doc.json` with interactive documentation at `/docs/`. Auto-generate clients in any language.
|===

== Credentials

Three credentials are involved in an on-premises deployment. They are distinct and serve different purposes.

[cols="1,1,2,1",options="header"]
|===
|Credential |Where it lives |What it does |Required?

|`LICENSE_KEY`
|AI service container (environment variable)
|Activates the AI service. A long string provided by the Tiny account representative.
|Yes — the service refuses to start without it.

|`TINYMCE_API_KEY`
|Editor page (CDN script URL) or build configuration
|Authenticates against `cdn.tiny.cloud` when loading TinyMCE from the CDN. This is the short string from the tiny.cloud dashboard.
|Only when loading TinyMCE from the CDN. Omit for self-hosted editor bundles.

|`license_key` (init option)
|`tinymce.init({ license_key: 'T8LK:...' })`
|Activates premium TinyMCE features when using a self-hosted editor bundle (not the CDN).
|Only for self-hosted editor deployments. Provided by the Tiny account representative.
|===

> NOTE: `LICENSE_KEY` (the AI service license) and `TINYMCE_API_KEY` / `license_key` (the editor license) are different credentials from different sources. Do not interchange them.

Reviewer 2: ```suggestion
NOTE: `LICENSE_KEY` (the AI service license) and `TINYMCE_API_KEY` / `license_key` (the editor license) are different credentials from different sources. They are not interchangable.
```

Author: Applied — aligned with your suggestion (fixed the typo to "interchangeable"). See 29e0a1c.

|Requirement |Details

|Container runtime
|Docker 20.10{plus}, Podman 4{plus}, or any OCI-compatible runtime. Kubernetes, AWS ECS, or Azure Container Apps are also supported.

|SQL database
|MySQL 8.0 or PostgreSQL 13{plus} (16 recommended).

|Redis
|3.2.6{plus} (7.x recommended). Single node or Cluster mode supported. Sentinel not supported.

|LLM access
|At least one provider. Multiple providers can coexist.

|License key and registry credentials
|Provided by a Tiny account representative.

|Token endpoint
|A back end that signs HS256 JWTs.

|Reverse proxy (recommended)
|The AI service does not terminate Transport Layer Security (TLS). A reverse proxy such as nginx, HAProxy, or a cloud load balancer is recommended for TLS termination in production.
|===

== Choosing a setup path

For evaluation, start with the xref:tinymceai-on-premises-getting-started.adoc[Getting started] guide (Docker Compose on a single machine). For production, work through each topic guide in order — the same components apply regardless of deployment target (Kubernetes, ECS, or standalone Docker).

[.text-center]
image::tinymceai-on-premises/complete-guide-fig-2.svg[Setup path decision tree,width=100%]

> == Topic guides

Reviewer 2: Do the topics mean to match each path on the flow chart?

Author: They're related but not 1:1. The flowchart shows deployment paths (evaluation vs production orchestrators), while the topic guides cover the full configuration lifecycle regardless of deployment path. The nav ordering (Database → Providers → JWT → Frameworks → Production) is the recommended reading sequence for all paths.
